core/src/apps/cardano/sign_tx.py

```diff
@@ -171,10 +171,11 @@ def _serialize_tx(keychain: seed.Keychain, msg: CardanoSignTx) -> Tuple[bytes, b
         msg.protocol_magic,
     )
 
-    # We always set transaction metadata to None, even if metadata
-    # hash is set. Metadata aren't sent to Trezor and the None
-    # should be replaced by the SW wallet if metadata exist.
-    serialized_tx = cbor.encode([tx_body, witnesses, None])
+    metadata = None
+    if msg.metadata:
+        metadata = cbor.Raw(bytes(msg.metadata))
```

Comment:
this opens up another CBOR injection
we'll need to decode the metadata first and verify that it's just one item
(personally i'd like to re-encode them afterwards, but i don't think the encoding is fully roundtrippable?)

Reply:
perhaps that should be part of the `Raw` class?
or call it something slightly different, something like "pre-encoded"? `EncodedItem`? so that it's clear that it is a single item but it is already encoded and included as-is, and at the same time safe to include without injections

Reply:
Does it? Since it's the last field of the transaction, can something be injected? -- I'll also check with the team once more.

Reply:
at the very least `metadata` could decode to more than one list item, breaking the tx format.
It _probably_ isn't security relevant, but who knows. Why even have injections.

Reply:
Validation added [here](https://github.com/trezor/trezor-firmware/pull/1145/commits/849c0710837c96f989aba391ee41dce683f599af). I think this should be sufficient. If it's invalid CBOR decoding fails, if there's some extra data at the end - decoding fails and if it's anything else than a `dict` after decoding validation also fails.